### PR TITLE
Clustered test runner for replicated loglet

### PR DESCRIPTION
**Pull request by jackkleeman**
Creates a test harness in the style of the loglet test harnesses, but that runs a replicated cluster using the local cluster runner to serve the loglet.

### CODE REVIEW

File: server/tests/common/replicated_loglet.rs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        replication,
+        nodeset: (1..log_server_count).collect(),
+        write_set: None,
+    };
+    let loglet_params = loglet_params.serialize()?;
+    base_config.bifrost.default_provider = ProviderKind::Replicated;
+    base_config.bifrost.default_provider_config = Some(loglet_params.clone());
+    base_config
+        .common
+        .set_bootstrap_num_partitions(NonZeroU16::new(1).unwrap());
```
Comment (AhmedSoliman) [inline server/tests/common/replicated_loglet.rs:114]: This is irrelevant to this environment, no?

File: server/tests/common/replicated_loglet.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+) -> googletest::Result<()>
+where
+    F: FnMut(TestEnv) -> O,
+    O: std::future::Future<Output = googletest::Result<()>> + Send,
+{
+    let loglet_params = ReplicatedLogletParams {
+        loglet_id: ReplicatedLogletId::new(1),
+        sequencer,
+        replication,
+        nodeset: (1..log_server_count).collect(),
```
Comment (AhmedSoliman) [inline server/tests/common/replicated_loglet.rs:106]: Should this be `(1..=log_server_count)`? I'm assuming that you want all nodes in the nodeset.

File: server/tests/common/replicated_loglet.rs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    );
+
+    let tc = TaskCenterBuilder::default()
+        .default_runtime_handle(tokio::runtime::Handle::current())
+        .ingress_runtime_handle(tokio::runtime::Handle::current())
+        .build()
+        .expect("task_center builds");
+
+    // ensure base dir lives longer than the node, otherwise it sees shutdown errors
+    // this will still respect LOCAL_CLUSTER_RUNNER_RETAIN_TEMPDIR=true
```
Comment (AhmedSoliman) [inline server/tests/common/replicated_loglet.rs:130]: Q: Not sure if the cluster utility already does that or not, but should we print on stdout this base_dir path after the test is done (or before)?

**Review by AhmedSoliman — APPROVED** [review]
Amazing work @jackkleeman. I love how the new utility makes writing multi-node tests a joy to use. I left a few minor comments but in general it's great!

File: server/tests/common/replicated_loglet.rs (hunk 3 above)
Comment (jackkleeman) [reply server/tests/common/replicated_loglet.rs:130]: it does indeed already do it

File: server/tests/common/replicated_loglet.rs (hunk 2 above)
Comment (jackkleeman) [reply server/tests/common/replicated_loglet.rs:106]: hmm. you make an interesting point

File: crates/local-cluster-runner/src/node/mod.rs [hunk 4]
```diff
@@ -150,6 +151,7 @@ impl Node {
         {
             let mut base_config = base_config.clone();
             base_config.common.allow_bootstrap = true;
+            base_config.common.force_node_id = Some(PlainNodeId::new(0));
```
[the remark this answers is not in the document]
Comment (jackkleeman) [reply crates/local-cluster-runner/src/node/mod.rs:154]: nope, nothing special

File: crates/local-cluster-runner/src/node/mod.rs [hunk 5]
```diff
@@ -679,6 +682,59 @@ impl StartedNode {
...
+
+        let nodes_config = metadata_client
+            .get::<NodesConfiguration>(NODES_CONFIG_KEY.clone())
+            .await;
+
+        let nodes_config = if let Ok(Some(nodes_config)) = nodes_config {
+            nodes_config
+        } else {
+            return false;
+        };
```
[the remark this answers is not in the document]
Comment (jackkleeman) [reply crates/local-cluster-runner/src/node/mod.rs:701]: whoa, i did not actually know about this

**Pull request merged**